File: src/SlidingWindow/FruitIntoBaskets.java
package SlidingWindow;

import java.util.HashMap;
import java.util.Map;

// https://leetcode.com/problems/fruit-into-baskets/
public class FruitIntoBaskets {
    public int totalFruit(int[] tree) {
        Map<Integer, Integer> count = new HashMap<>();
        int res = 0;
        int i = 0, j;
        for (j = 0; j < tree.length; ++j) {
            count.put(tree[j], count.getOrDefault(tree[j], 0) + 1);
            while (count.size() > 2) {
                count.put(tree[i], count.get(tree[i]) - 1);
                count.remove(tree[i++], 0);
            }
            res = Math.max(res, j - i + 1);
        }
        return res;
    }

    public static void main(String[] args) {
        FruitIntoBaskets f = new FruitIntoBaskets();
        int[] input = {3,3,3,1,2,1,1,2,3,3,4};
        System.out.println("Max fruits: " + f.totalFruit(input));
    }
}

File: py/FruitIntoBaskets.py
# https://leetcode.com/problems/fruit-into-baskets/description/

from collections import defaultdict
from typing import List


class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        typeCount, left = defaultdict(int), 0

        for right, _ in enumerate(fruits):
            typeCount[fruits[right]] += 1
            if len(typeCount) > 2:
                typeCount[fruits[left]] -= 1
                if typeCount[fruits[left]] == 0:
                    del typeCount[fruits[left]]
                left += 1
        return right - left + 1


s = Solution()
print(s.totalFruit([1, 2, 3, 2, 2]))
